src/plugins/hiddenColumns/contextMenuItem/showColumn.js

```diff
@@ -6,112 +6,100 @@ import * as C from '../../../i18n/constants';
-          hiddenColumnsPlugin.showColumns(afterHiddenColumns);
-          afterHiddenColumns.length = 0;
-        }
+      const [, startVisualColumn, , endVisualColumn] = this.getSelectedLast();
+      const onlyFirstVisibleColumnSelected =
+        this.columnIndexMapper.getRenderableFromVisualIndex(startVisualColumn) === 0 &&
+        startVisualColumn === endVisualColumn;
+      const onlyLastVisibleColumnSelected =
+        this.columnIndexMapper.getRenderableFromVisualIndex(endVisualColumn) === this.countRenderableColumns() - 1 &&
+        startVisualColumn === endVisualColumn;
+
+      let startPhysicalColumn = this.toPhysicalColumn(startVisualColumn);
+      let endPhysicalColumn = this.toPhysicalColumn(endVisualColumn);
+
+      if (onlyFirstVisibleColumnSelected) {
+        startPhysicalColumn = 0;
+      }
 
-      } else {
-        rangeEach(start, end, column => hiddenColumnsPlugin.showColumn(column));
+      if (onlyLastVisibleColumnSelected) {
+        endPhysicalColumn = this.countSourceCols() - 1;
       }
 
+      hiddenColumnsPlugin.showColumns(columns);
+
+      const startVisualColumnAfterAction = this.toVisualColumn(startPhysicalColumn);
+      const endVisualColumnAfterAction = this.toVisualColumn(endPhysicalColumn);
+
+      // Selection start and selection end coordinates might be changed after showing some items.
+      this.selectColumns(startVisualColumnAfterAction, endVisualColumnAfterAction);
       this.render();
       this.view.wt.wtOverlays.adjustElementsSize(true);
     },
     disabled: false,
     hidden() {
-      if (!hiddenColumnsPlugin.hiddenColumns.length || !this.selection.isSelectedByColumnHeader()) {
+      if (!this.selection.isSelectedByColumnHeader() || hiddenColumnsPlugin.getHiddenColumns().length < 1) {
         return true;
       }
 
-      beforeHiddenColumns.length = 0;
-      afterHiddenColumns.length = 0;
+      columns.length = 0;
 
-      const { from, to } = this.getSelectedRangeLast();
-      const start = Math.min(from.col, to.col);
-      const end = Math.max(from.col, to.col);
-      let hiddenInSelection = false;
+      const [, startColumn, , endColumn] = this.getSelectedLast();
+      const start = Math.min(startColumn, endColumn);
+      const end = Math.max(startColumn, endColumn);
+      const sequence = this.columnIndexMapper.getIndexesSequence();
+      const physicalStartColumn = this.columnIndexMapper.getPhysicalFromVisualIndex(start);
+      const currentStartPosition = sequence.indexOf(physicalStartColumn);
 
       if (start === end) {
-        let totalColumnLength = this.countSourceCols();
+        const columnsBefore = [];
+        const columnsAfter = [];
 
-        rangeEach(0, totalColumnLength, (column) => {
-          const partedHiddenLength = beforeHiddenColumns.length + afterHiddenColumns.length;
+        rangeEach(0, currentStartPosition - 1, (column) => {
+          if (hiddenColumnsPlugin.isHidden(sequence[column], true)) {
```

Comment:
`isHidden` is used with second parameter, this is obsolete and not used in hiddenColumns plugin. We should double-check if this works as expected.
```suggestion
          if (hiddenColumnsPlugin.isHidden(sequence[column])) {
```

Reply:
I refactored the whole function within https://github.com/handsontable/handsontable/commit/712d62b6a23582b2ddbd83c4ea720c458e5f85e3#diff-3e0e80185f255ec40def24664b448366. I hope it's more clear. I'm looking forward to your feedback.

Reply:
Looks a lot cleaner now! 👌